File: imgcat.go
package imgcat

import (
	"bufio"
	"bytes"
	"encoding/base64"
	"fmt"
	"os"
	"strings"
)

type ImageCat struct {
	bf bytes.Buffer
}

func (i *ImageCat) Write(p []byte) (n int, err error) {
	return i.bf.Write(p)
}

func (i *ImageCat) Cat() {
	TERM := os.Getenv("TERM")
	encoded := base64.StdEncoding.EncodeToString(i.bf.Bytes())

	_IMG_PRE := "\033]"
	_IMG_POST := "\a"

	if !strings.HasPrefix(TERM, "screen") {
		_IMG_PRE = "\033Ptmux;\033\033]"
		_IMG_POST = "\a\033\\\n"
	}
	inline := 1
	preserveAspectRatio := 0
	data := fmt.Sprintf("\n%s1337;File=inline=%d;preserveAspectRatio=%d:%s%s",
		_IMG_PRE, inline, preserveAspectRatio, encoded, _IMG_POST)
	fmt.Println("=================")
	bio := bufio.NewWriter(os.Stdout)
	bio.WriteString(data)
	bio.Flush()

	return
}

File: imgcat.py
#!/usr/bin/python

import os
import base64
import codecs
import sys

TERM = os.environ.get('TERM')
TERM_IS_SCREEN = TERM and TERM.startswith('screen')

_IMG_PRE = '\033Ptmux;\033\033]' if TERM_IS_SCREEN else '\033]'
_IMG_POST = '\a\033\\' if TERM_IS_SCREEN else '\a'


def _read_as_base64(path):
    with codecs.open(path, mode='rb') as fh:
        return base64.b64encode(fh.read())


def imgcat(path, inline=1, preserve_aspect_ratio=0, **kwargs):
    return '\n%s1337;File=inline=%d;preserveAspectRatio=%d:%s%s' % (
        _IMG_PRE, inline, preserve_aspect_ratio,
        _read_as_base64(path), _IMG_POST)

if __name__ == '__main__':
    if len(sys.argv)<2:
        exit()
    print imgcat(sys.argv[1])
